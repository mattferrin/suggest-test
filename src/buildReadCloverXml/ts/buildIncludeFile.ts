import { MeowOutput } from "../../bin/ts/cli";
import { SummaryItem } from "./undefinedOrSummary";

export function buildIncludeFile(input: MeowOutput) {
  return function includeFile(file: SummaryItem | undefined) {
    const isTested =
      (file?.statements ?? 0) > (file?.uncoveredStatements ?? 0) ||
      (file?.conditionals ?? 0) > (file?.uncoveredConditionals ?? 0);
    const isIncludeTested = input.flags.includeTested && isTested;

    const isNotTested =
      (file?.statements ?? 0) === (file?.uncoveredStatements ?? 0) &&
      (file?.conditionals ?? 0) === (file?.uncoveredConditionals ?? 0);
    const isIncludeNotTested = input.flags.includeNotTested && isNotTested;

    return isIncludeTested || isIncludeNotTested;
  };
}
