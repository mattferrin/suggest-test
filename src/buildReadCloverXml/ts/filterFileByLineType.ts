import { MeowOutput } from "../../bin/ts/cli";
import { FileArray } from "./summarizedResult";
import { Untrusted } from "./Untrusted";

export function filterFileByLineType(
  input: MeowOutput,
  projectPackage: Untrusted<{
    readonly file: FileArray;
  }>
): Untrusted<FileArray> {
  if (!input.flags.includeStatements) {
    return projectPackage?.file?.filter((projectPackageFile) => {
      return projectPackageFile?.line?.some((line) => {
        return line?.$?.type === "cond";
      });
    });
  } else {
    return projectPackage?.file;
  }
}
