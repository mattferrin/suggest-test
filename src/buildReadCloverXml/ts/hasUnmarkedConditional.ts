import { MeowOutput } from "../../bin/ts/cli";
import { FileArray } from "./summarizedResult";
import { Untrusted } from "./Untrusted";

export function hasUnmarkedConditional(
  input: MeowOutput,
  file: Untrusted<FileArray[number]>
): boolean {
  if (input.flags.includeConditionals) {
    const hasConditional =
      file?.metrics?.some((metric) => {
        return Number.parseInt(metric?.$?.conditionals ?? "0") > 0;
      }) ?? false;

    return (
      hasConditional &&
      (file?.line?.every((line) => {
        return line?.$?.type !== "cond";
      }) ??
        false)
    );
  } else {
    return false;
  }
}
