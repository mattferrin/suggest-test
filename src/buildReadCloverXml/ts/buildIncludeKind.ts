import { MeowOutput } from "../../bin/ts/cli";
import { SummaryItem } from "./undefinedOrSummary";

export function buildIncludeKind(input: MeowOutput) {
  return function includeKind(file: SummaryItem | undefined) {
    if (file === undefined) {
      return false;
    } else {
      const isStatementKeeper =
        (file.statements ?? 0) > 0 && file.uncoveredStatements > 0;
      const isConditionalKeeper =
        (file.conditionals ?? 0) > 0 && file.uncoveredConditionals > 0;

      if (!input.flags.includeConditionals) {
        return isStatementKeeper;
      } else if (!input.flags.includeStatements) {
        return isConditionalKeeper;
      } else {
        return isStatementKeeper || isConditionalKeeper;
      }
    }
  };
}
