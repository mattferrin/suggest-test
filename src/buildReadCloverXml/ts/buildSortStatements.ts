import { MeowOutput } from "../../bin/ts/cli";
import { ReadCloverXmlOutput } from "./buildReadCloverXml";
import { summarizeIntoString } from "./summarizeIntoString";

export function buildSortStatements(input: MeowOutput) {
  return function sortStatements(
    output: ReadCloverXmlOutput
  ): ReadCloverXmlOutput {
    switch (output.tag) {
      case "errors":
        return output;
      case "summary": {
        const summaryArray = output.summaryArray
          ?.filter((file) => (file?.statements ?? 0) > 1)
          ?.sort((fileA, fileB) => {
            if (
              fileB?.statements === null ||
              fileB?.statements === undefined ||
              fileA?.statements === null ||
              fileA?.statements === undefined
            ) {
              return -1;
            } else {
              return fileA.weightedStatements > fileB.weightedStatements
                ? -1
                : 1;
            }
          });

        return {
          tag: "summary",
          summaryArray,
          summaryString: summarizeIntoString(summaryArray),
        };
      }
    }
  };
}
