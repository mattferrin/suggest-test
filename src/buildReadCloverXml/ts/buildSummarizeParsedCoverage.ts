import { MeowOutput } from "../../bin/ts/cli";
import { filterFileByLineType } from "./filterFileByLineType";
import { FileArray, summarizedResult } from "./summarizedResult";
import { SummaryItem } from "./undefinedOrSummary";
import { Untrusted } from "./Untrusted";

export type Parsed = Untrusted<{
  readonly coverage: {
    readonly project: ReadonlyArray<{
      readonly package: ReadonlyArray<{
        readonly file: FileArray;
      }>;
    }>;
  };
}>;

type SummarizeParsedCoverageOutput =
  | ReadonlyArray<SummaryItem | undefined>
  | undefined;

export function buildSummarizeParsedCoverage(input: MeowOutput) {
  return function summarizeParsedCoverage(
    parsed: Parsed
  ): SummarizeParsedCoverageOutput {
    return parsed?.coverage?.project?.flatMap((project) => {
      return project?.package?.flatMap((projectPackage) => {
        const filteredProjectPackageFile = filterFileByLineType(
          input,
          projectPackage
        );

        return summarizedResult(input, filteredProjectPackageFile);
      });
    });
  };
}
