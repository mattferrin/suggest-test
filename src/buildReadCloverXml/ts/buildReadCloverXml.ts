import fs from "fs/promises";
import { parseStringPromise } from "xml2js";
import { MeowOutput } from "../../bin/ts/cli";
import { buildSortStatements } from "./buildSortStatements";
import { buildSummarizeParsedCoverage } from "./buildSummarizeParsedCoverage";
import { passThroughLogger } from "./passThroughLogger";
import { SummaryItem, undefinedOrSummary } from "./undefinedOrSummary";

export type ReadCloverXmlOutput =
  | { readonly tag: "errors"; readonly errors: readonly Error[] }
  | {
      readonly tag: "summary";
      readonly summaryArray: ReadonlyArray<SummaryItem | undefined>;
      readonly summaryString: string | { readonly tag: "not-ready" };
    };

export function buildReadCloverXml(input: MeowOutput) {
  // eslint-disable-next-line functional/functional-parameters
  return async function readCloverXml(): Promise<ReadCloverXmlOutput> {
    return await fs
      .readFile(input.input[0] ?? "coverage/clover.xml")
      .then(parseStringPromise)
      .then(buildSummarizeParsedCoverage(input))
      .then(undefinedOrSummary)
      .then(buildSortStatements(input))
      .catch((reason): ReadCloverXmlOutput => {
        return { tag: "errors", errors: [new Error(reason)] };
      })
      .then(passThroughLogger);
  };
}
