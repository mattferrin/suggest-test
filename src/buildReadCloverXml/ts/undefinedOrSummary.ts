import { ReadCloverXmlOutput } from "./buildReadCloverXml";

export interface SummaryItem {
  readonly conditionals: number | null | undefined;
  readonly path: string | null | undefined;
  readonly statements: number | null | undefined;
  readonly uncoveredConditionals: number;
  readonly uncoveredStatements: number;
  readonly score: number;
  readonly hasUnmarkedConditional: boolean;
}

export function undefinedOrSummary(
  summary: ReadonlyArray<SummaryItem | undefined> | undefined
): ReadCloverXmlOutput {
  if (summary === undefined || summary === null) {
    return {
      tag: "errors",
      errors: [
        new Error(
          "The parsed clover.xml has an undefined or null property somewhere"
        ),
      ],
    };
  } else {
    return {
      tag: "summary",
      summaryArray: summary,
      summaryString: { tag: "not-ready" },
    };
  }
}
