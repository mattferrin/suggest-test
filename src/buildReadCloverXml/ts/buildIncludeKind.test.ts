import { MeowOutput } from "../../bin/ts/cli";
import { buildIncludeKind } from "./buildIncludeKind";
import { SummaryItem } from "./undefinedOrSummary";

function snap(
  uuid: string,
  comment: string | null,
  input: MeowOutput,
  file: SummaryItem | undefined
): void {
  it(uuid, () => {
    /** setup mocks */

    /** unit under test */
    const includeKind = buildIncludeKind(input);
    const result = includeKind(file);

    /** assertions */
    expect({
      [`${comment ?? "TODO: comment"}`]: { result },
    }).toMatchSnapshot();
  });
}

snap(
  "f2c0498a-c3cc-4a41-80fd-f588e0a99235",
  "returns `false` to exclude undefined file item WHEN includeConditionals false",
  { flags: { includeConditionals: false } } as MeowOutput,
  undefined
);

snap(
  "528059d4-1a4b-4cab-af69-c7ab0418506c",
  "returns `false` to exclude undefined file item WHEN includeStatements false",
  {
    flags: { includeStatements: false },
  } as MeowOutput,
  undefined
);

snap(
  "5c8f57e5-9bec-4908-94bb-0e4d5cd19981",
  "returns `false` to exclude undefined file item WHEN both includeConditionals AND includeStatements true",
  {
    flags: { includeConditionals: true, includeStatements: true },
  } as MeowOutput,
  undefined
);

snap(
  "7cf788e5-cee9-44f7-baa0-1af2e17de2ab",
  "returns `true` to include item WHEN 1 uncovered statement AND conditionals excluded",
  {
    flags: { includeConditionals: false, includeStatements: true },
  } as MeowOutput,
  { statements: 1, uncoveredStatements: 1 } as SummaryItem
);

snap(
  "17dc0f2d-3ca0-43c5-87b7-1d2aa236dee9",
  "returns `true` to include item WHEN 1 uncovered conditional AND statements excluded",
  {
    flags: { includeConditionals: true, includeStatements: false },
  } as MeowOutput,
  { conditionals: 1, uncoveredConditionals: 1 } as SummaryItem
);

snap(
  "e4177bfa-d1a6-4953-a625-144726b42906",
  "returns `true` to include item WHEN 1 uncovered statement AND all included",
  {
    flags: { includeConditionals: true, includeStatements: true },
  } as MeowOutput,
  { statements: 1, uncoveredStatements: 1 } as SummaryItem
);

snap(
  "2139f6fd-719c-424d-b4b2-bfbb2053c497",
  "returns `true` to include item WHEN 1 uncovered conditional AND all included",
  {
    flags: { includeConditionals: true, includeStatements: true },
  } as MeowOutput,
  { conditionals: 1, uncoveredConditionals: 1 } as SummaryItem
);

snap(
  "cad6ea7f-4248-433d-989b-0f7c6d403e25",
  "returns `false` to exclude item WHEN 0 statements AND conditionals excluded",
  {
    flags: { includeConditionals: false, includeStatements: true },
  } as MeowOutput,
  { statements: 0, uncoveredStatements: 1 } as SummaryItem
);

snap(
  "71b9c9f8-d795-4234-87ca-a2da7ddf7489",
  "returns `false` to exclude item WHEN 0 conditionals AND statements excluded",
  {
    flags: { includeConditionals: true, includeStatements: false },
  } as MeowOutput,
  { conditionals: 0, uncoveredConditionals: 1 } as SummaryItem
);

snap(
  "4bf9a8fc-eaa4-412f-a7ac-81fc4d82b3c2",
  "returns `false` to exclude item WHEN 0 conditionals AND 0 statements AND all included",
  {
    flags: { includeConditionals: true, includeStatements: true },
  } as MeowOutput,
  {
    conditionals: 0,
    uncoveredConditionals: 1,
    statements: 0,
    uncoveredStatements: 1,
  } as SummaryItem
);

snap(
  "cfe960a0-28f2-42d4-ac23-e303109a30b4",
  "returns `false` to exclude item WHEN 1 statement that is covered AND conditionals excluded",
  {
    flags: { includeConditionals: false, includeStatements: true },
  } as MeowOutput,
  { statements: 1, uncoveredStatements: 0 } as SummaryItem
);

snap(
  "07e7baf7-9408-4254-af30-a7c724b4b7ad",
  "returns `false` to exclude item WHEN 1 conditional that is covered AND statements excluded",
  {
    flags: { includeConditionals: true, includeStatements: false },
  } as MeowOutput,
  { conditionals: 1, uncoveredConditionals: 0 } as SummaryItem
);

snap(
  "96dc42c6-bade-45fb-aa2c-57d9baf3367c",
  "returns `false` to exclude item WHEN 1 conditional that is covered AND 1 statement that is covered AND all included",
  {
    flags: { includeConditionals: true, includeStatements: true },
  } as MeowOutput,
  {
    conditionals: 1,
    uncoveredConditionals: 0,
    statements: 1,
    uncoveredStatements: 0,
  } as SummaryItem
);
