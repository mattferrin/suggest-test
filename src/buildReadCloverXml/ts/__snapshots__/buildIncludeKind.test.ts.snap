// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`07e7baf7-9408-4254-af30-a7c724b4b7ad 1`] = `
{
  "returns \`false\` to exclude item WHEN 1 conditional that is covered AND statements excluded": {
    "result": false,
  },
}
`;

exports[`4bf9a8fc-eaa4-412f-a7ac-81fc4d82b3c2 1`] = `
{
  "returns \`false\` to exclude item WHEN 0 conditionals AND 0 statements AND all included": {
    "result": false,
  },
}
`;

exports[`5c8f57e5-9bec-4908-94bb-0e4d5cd19981 1`] = `
{
  "returns \`false\` to exclude undefined file item WHEN both includeConditionals AND includeStatements true": {
    "result": false,
  },
}
`;

exports[`7cf788e5-cee9-44f7-baa0-1af2e17de2ab 1`] = `
{
  "returns \`true\` to include item WHEN 1 uncovered statement AND conditionals excluded": {
    "result": true,
  },
}
`;

exports[`17dc0f2d-3ca0-43c5-87b7-1d2aa236dee9 1`] = `
{
  "returns \`true\` to include item WHEN 1 uncovered conditional AND statements excluded": {
    "result": true,
  },
}
`;

exports[`71b9c9f8-d795-4234-87ca-a2da7ddf7489 1`] = `
{
  "returns \`false\` to exclude item WHEN 0 conditionals AND statements excluded": {
    "result": false,
  },
}
`;

exports[`96dc42c6-bade-45fb-aa2c-57d9baf3367c 1`] = `
{
  "returns \`false\` to exclude item WHEN 1 conditional that is covered AND 1 statement that is covered AND all included": {
    "result": false,
  },
}
`;

exports[`2139f6fd-719c-424d-b4b2-bfbb2053c497 1`] = `
{
  "returns \`true\` to include item WHEN 1 uncovered conditional AND all included": {
    "result": true,
  },
}
`;

exports[`528059d4-1a4b-4cab-af69-c7ab0418506c 1`] = `
{
  "returns \`false\` to exclude undefined file item WHEN includeStatements false": {
    "result": false,
  },
}
`;

exports[`cad6ea7f-4248-433d-989b-0f7c6d403e25 1`] = `
{
  "returns \`false\` to exclude item WHEN 0 statements AND conditionals excluded": {
    "result": false,
  },
}
`;

exports[`cfe960a0-28f2-42d4-ac23-e303109a30b4 1`] = `
{
  "returns \`false\` to exclude item WHEN 1 statement that is covered AND conditionals excluded": {
    "result": false,
  },
}
`;

exports[`e4177bfa-d1a6-4953-a625-144726b42906 1`] = `
{
  "returns \`true\` to include item WHEN 1 uncovered statement AND all included": {
    "result": true,
  },
}
`;

exports[`f2c0498a-c3cc-4a41-80fd-f588e0a99235 1`] = `
{
  "returns \`false\` to exclude undefined file item WHEN includeConditionals false": {
    "result": false,
  },
}
`;
